### DISCUSSION TITLE
Proper way to access files in `/public` from server-side functions from Remix?

### DISCUSSION THREAD

**Discussion by colmanhumphrey** [vercel/community#1305] [Help]
I'm using Remix. I have a text file in `/public`, e.g. `/public/some_file.txt`. I'm able to use it in `someutil.server.ts` within `/app` by pulling it in as `readFileSync("./public/some_file.txt")` locally, but this fails on Vercel, (`ENOENT: no such file or directory` and so on).

I finally cobbled together a solution using ``readFileSync(`${__dirname}/../../public/some_file.txt`)`` based on reading a few issues here and there, but it feels a little bit brittle (and I can't see much documentation, e.g. https://vercel.com/guides/how-can-i-use-files-in-serverless-functions doesn't seem to work), and I assume there's some server-side parsing of this because I can't see anything in the relevant folders if I just log the files there (and `path.join` and `process.cwd()` didn't seem to work).

Is there a more correct way to do this? Or if this is right, did I just miss some docs somewhere?

Thanks!

**Comment by colmanhumphrey** [comment]
Ahh @theMosaad has pointed me in the right direction: see https://mobile.twitter.com/rauchg/status/1548732499680845825 from @rauchg and https://www.themosaad.com/blog/loading-static-file-remix-vercel.

This I think helps explain the following strange thing I saw. I used the following:

```ts
const isVercel = process.env.VERCEL === "1" || false
const filePath = isVercel ? `${__dirname}/../../public/some_file.txt` : "./public/some_file.txt"
...
// later, something with `readFileSync(filePath...`
```

And this DOES NOT work. But replacing `"./public/some_file.txt"` with `"/public/some_file.txt"` does work (remove the `.`, in fact this of course breaks locally), even though in theory on Vercel itself, neither value is relevant: they're not used, and I guess the location isn't present on the server?

Thankfully I then realised that `${__dirname}/../../public/some_file.txt` works locally too, so I could remove all the `isVercel` stuff. But weird.

**Reply by colmanhumphrey** [reply]
note that I'm unmarking this as the answer because this isn't an answer, it's just added context for my initial question. An answer would address:
- is there a better/more correct way to do this?
- if not, should this be better documented?

**Comment by colmanhumphrey** [comment]
OK this apparently broke again. Has this changed in the last month?

**Comment by WillCalderwood** [comment]
This seems to do the job assuming your original path is something like `"./public/files/filename.txt"`

This works both locally and on Vercel.
 ```
import path from "node:path";

export class FileUtils {
  /**
   * Retrieves the public file path for a given original path.
   * We need to do this for Vercel, otherwise it won't find the file.
   *
   * @param {string} originalPath - The original path of the file.
   * @return {string} The resolved public file path.
   */
  public static getPublicFilePath(originalPath: string): string {
    originalPath = originalPath.replace("./public/", "");
    return path.resolve('./public', originalPath);
  }
}
```

**Reply by Wolemercy** [reply]
This worked like a charm. Thank youu!
